tools/kvmexit_example.txt

```diff
@@ -0,0 +1,106 @@
+Demonstrations of kvm exit reasons, the Linux eBPF/bcc version.
+
+
+This tool displays the detail exit_reason and its statistics of each
+kvm_exit for all virtual machines running on this physical machine.
```

Comment:
The above explanation only shows what you intended to do. It would be good to explain how
such information will help you solve any particular issues. You mentioned you have many VMs and performance is important, how many VMs do you have in each physical box?

For example, I tried on simple qemu/kvm run (start and at login prompt, quit), I got:
```
$ sudo python3.6 kvmexit.py
Display kvm exit reasons and statistics... Hit Ctrl-C to end.
TGID     PID      KVM_EXIT_REASON                     STAT
^C4036993  4037007  EXIT_REASON_EXTERNAL_INTERRUPT      728     
4036993  4037007  EXIT_REASON_CPUID                   196     
4036993  4037007  EXIT_REASON_HLT                     1447    
4036993  4037007  EXIT_REASON_VMCALL                  424     
4036993  4037007  EXIT_REASON_CR_ACCESS               11      
4036993  4037007  EXIT_REASON_DR_ACCESS               1       
4036993  4037007  EXIT_REASON_IO_INSTRUCTION          19642   
4036993  4037007  EXIT_REASON_MSR_READ                92      
4036993  4037007  EXIT_REASON_MSR_WRITE               1989    
4036993  4037007  EXIT_REASON_PAUSE_INSTRUCTION       666     
4036993  4037007  EXIT_REASON_GDTR_IDTR               6       
4036993  4037007  EXIT_REASON_LDTR_TR                 3       
4036993  4037007  EXIT_REASON_EPT_VIOLATION           2658    
4036993  4037007  EXIT_REASON_PREEMPTION_TIMER        690     
4036993  4037007  EXIT_REASON_WBINVD                  1       
4036993  4037007  EXIT_REASON_XSETBV                  1       
4036993  4037007  EXIT_REASON_APIC_WRITE              1182
```
How does the above information helps you?

You could you use debugfs kvm/kvm_exit tracepoint to get the same information, I guess the reason might be due to performance reason so you want to do in-kernel aggregation. So clarifying your use case is important here.

Reply:
> The above explanation only shows what you intended to do. It would be good to explain how
> such information will help you solve any particular issues. You mentioned you have many VMs and performance is important, how many VMs do you have in each physical box?
> 
The mainly purpose of doing this is for performance optimization in the virtual machine scenario. As we know, vms'  frequent exits can cause performance problems, thus we want to locate the frequent exited reasons and then find solutions to reduce or avoid the exit. 

> For example, I tried on simple qemu/kvm run (start and at login prompt, quit), I got:
> 
> ```
> $ sudo python3.6 kvmexit.py
> Display kvm exit reasons and statistics... Hit Ctrl-C to end.
> TGID     PID      KVM_EXIT_REASON                     STAT
> ^C4036993  4037007  EXIT_REASON_EXTERNAL_INTERRUPT      728     
> 4036993  4037007  EXIT_REASON_CPUID                   196     
> 4036993  4037007  EXIT_REASON_HLT                     1447    
> 4036993  4037007  EXIT_REASON_VMCALL                  424     
> 4036993  4037007  EXIT_REASON_CR_ACCESS               11      
> 4036993  4037007  EXIT_REASON_DR_ACCESS               1       
> 4036993  4037007  EXIT_REASON_IO_INSTRUCTION          19642   
> 4036993  4037007  EXIT_REASON_MSR_READ                92      
> 4036993  4037007  EXIT_REASON_MSR_WRITE               1989    
> 4036993  4037007  EXIT_REASON_PAUSE_INSTRUCTION       666     
> 4036993  4037007  EXIT_REASON_GDTR_IDTR               6       
> 4036993  4037007  EXIT_REASON_LDTR_TR                 3       
> 4036993  4037007  EXIT_REASON_EPT_VIOLATION           2658    
> 4036993  4037007  EXIT_REASON_PREEMPTION_TIMER        690     
> 4036993  4037007  EXIT_REASON_WBINVD                  1       
> 4036993  4037007  EXIT_REASON_XSETBV                  1       
> 4036993  4037007  EXIT_REASON_APIC_WRITE              1182
> ```
> 
> How does the above information helps you?
> 
> You could you use debugfs kvm/kvm_exit tracepoint to get the same information, I guess the reason might be due to performance reason so you want to do in-kernel aggregation. So clarifying your use case is important here.

Yes, using bcc/ebpf is for performance considerations.  Thanks for the kindly reminder, I will do the polish. And then push a new version for all modifications later. :)

Reply:
Linux only supports several exit reasons 'exits  halt_exits  io_exits  irq_exits  mmio_exits  signal_exits' for a virtual machine, actually we need to know the full reasons especially wrmsr/rdmsr/ept-violation/ple on a x86 server. This could not be implemented in  upstream kernel(debugfs kvm) because it's x86-specific([discuss](https://patchwork.kernel.org/project/kvm/patch/1555939499-30854-1-git-send-email-pizhenwei@bytedance.com/) about this).

Reply:
All these are important reasons to have a separate tool where tracepoints cannot provide, e.g., 
   - tracepoint won't do in-kernel aggregation, and bpf program provides nice kernel aggregation, bpf in-kernel cache may further improve performance.
   - tracepoint missed some arch (e.g. x86) specific kvm reasons.
   - the detailed exit reason distribution may help tuning qemu and guest kernel to avoid hot vm exit.
It would be good if you can put some versions of the above reasoning in the beginning of the tool as the comments os it would be clear of why and what this tool can help.

Reply:
> All these are important reasons to have a separate tool where tracepoints cannot provide, e.g.,
> 
> * tracepoint won't do in-kernel aggregation, and bpf program provides nice kernel aggregation, bpf in-kernel cache may further improve performance.
> * tracepoint missed some arch (e.g. x86) specific kvm reasons.
> * the detailed exit reason distribution may help tuning qemu and guest kernel to avoid hot vm exit.
>   It would be good if you can put some versions of the above reasoning in the beginning of the tool as the comments os it would be clear of why and what this tool can help.

Hi @yonghong-song , thanks for your kindly suggestions, and I have prepared a version 2 based on v1, please help to review. Thanks a lot. :)